lib/Controller/HarpController.php

```diff
@@ -0,0 +1,165 @@
+<?php
+
+declare(strict_types=1);
+
+/**
+ * SPDX-FileCopyrightText: 2023 Nextcloud GmbH and Nextcloud contributors
+ * SPDX-License-Identifier: AGPL-3.0-or-later
+ */
+
+namespace OCA\AppAPI\Controller;
+
+use OCA\AppAPI\AppInfo\Application;
+use OCA\AppAPI\Service\DaemonConfigService;
+use OCA\AppAPI\Service\ExAppService;
+use OCA\AppAPI\Service\HarpService;
+use OCP\AppFramework\Controller;
+use OCP\AppFramework\Http;
+use OCP\AppFramework\Http\Attribute\NoCSRFRequired;
+use OCP\AppFramework\Http\Attribute\PublicPage;
+use OCP\AppFramework\Http\DataResponse;
+use OCP\IAppConfig;
+use OCP\IGroupManager;
+use OCP\IRequest;
+use OCP\IUserManager;
+use OCP\Security\Bruteforce\IThrottler;
+use OCP\Security\ICrypto;
+use Psr\Log\LoggerInterface;
+
+class HarpController extends Controller {
+	protected $request;
+
+	public function __construct(
+		IRequest                             $request,
+		private readonly IAppConfig          $appConfig,
+		private readonly ExAppService        $exAppService,
+		private readonly LoggerInterface     $logger,
+		private readonly IThrottler          $throttler,
+		private readonly IUserManager        $userManager,
+		private readonly IGroupManager       $groupManager,
+		private readonly DaemonConfigService $daemonConfigService,
+		private readonly ICrypto             $crypto,
+		private readonly ?string             $userId,
+	) {
+		parent::__construct(Application::APP_ID, $request);
+
+		$this->request = $request;
+	}
+
+	private function validateHarpSharedKey(string $appId): bool {
```

Comment:
should accept `ExApp` instead of `string`, imho.

In `getExAppMetadata` then first will be call to `getExApp` and this will be passed to `validateHarpSharedKey`

This will help to call `getExApp` only once instead of two times for the  fetching ExApp info from HaRP(one network call takes 3ms-20ms).

In `getUserInfo` just needed to add new the `getExApp` call, nothing will be changed except ~5 more lines of code total.

Reply:
nice!